msal/src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -266,9 +273,10 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
             }
         }
 
-        private bool ContainsClaimsRequiredToStoreInCache(IdToken idToken, RequestContext requestContext)
+        private bool ContainsClaimsRequiredToStoreInCache(IdToken idToken, RequestContext requestContext, Core.Instance.AuthorityType authorityType)
         {
-            if (idToken == null || string.IsNullOrEmpty(idToken.TenantId) || string.IsNullOrEmpty(idToken.PreferredUsername))
+            //PreferredUsername is not required in B2C cache scenarios
+            if (idToken == null || string.IsNullOrEmpty(idToken.TenantId) || (string.IsNullOrEmpty(idToken.PreferredUsername) && authorityType != Core.Instance.AuthorityType.B2C))
```

Comment:
Unless MSAL .NET uses preferred username for its legacy caching purposes, it shouldn't be required for caching at all.

Reply:
From my testing, when I create a token cache entry, the preferredUserName is not present in either the key or the value. However, this check will prevent a response from being cached if the preferedUserName is not set regardless. I am not sure if I am missing something.

@SomkaPe can you explain why there is a check for preferedUserName?

Reply:
@trwalke did you test this and other flows with-out the preferredUserName set?

Reply:
PreferredUsername has to be present in the Account object in the payload. But it's not used as part of the cache key. Therefore, it's not crucial for the caching purposes.

Reply:
Did you test with this change or before this change? What happens with this change. Did you validate where else the preferredUserName is used? And why?

---
In reply to: [215019267](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#discussion_r215019267) [](ancestors = 215019267)

Reply:
@oldalton The application fails to put the access token and refresh token into the cache when the id token had an empty PreferedUserName field. 

@henrik-me this was tested before and after my change. Whenever I set the preferredUserName to empty, the access tokens and refresh tokens were not put into the cache.

as for what the preferredUserName is used for, it is used in the iAccount objects and it is used in the TokenCacheNotificationArgs parameters that are used in the BeforeWrite, BeforeAccess and AfterAccess delegates, giving the user the ability to use the preferredUserName as they see fit

Reply:
Yes, this condition was removed specifically from the B2C case so that tokens are cached